packages/framework/src/Model/Product/Recalculation/ProductRecalculationRepository.php

```diff
@@ -21,57 +22,96 @@ public function __construct(
      * @param int[] $productIds
      * @return int[]
      */
-    public function getIdsToRecalculate(array $productIds): array
+    protected function dropVariantIdsFromProductIds(array $productIds): array
     {
-        $mainVariantsIds = $this->getMainVariantIdsOfVariants($productIds);
-
-        $productIdsWithAddedMainVariantIds = [...$productIds, ...$mainVariantsIds];
-
-        $regularProductOrMainVariantIds = $this->dropVariantIdsFromProductIds($productIdsWithAddedMainVariantIds);
-        $variantIds = $this->getVariantIds($productIdsWithAddedMainVariantIds);
+        $result = $this->em->createQuery('SELECT p.id FROM ' . Product::class . ' p WHERE p.id IN (:productIds) AND p.variantType != :variantTypeVariant')
+            ->setParameter('productIds', $productIds)
+            ->setParameter('variantTypeVariant', Product::VARIANT_TYPE_VARIANT)
+            ->getResult();
 
-        return [...$variantIds, ...$regularProductOrMainVariantIds];
+        return array_column($result, 'id');
     }
 
     /**
-     * @param int[] $productIds
-     * @return int[]
+     * @param int[] $mainVariantsIds
+     * @return int[][]
      */
-    protected function getMainVariantIdsOfVariants(array $productIds): array
+    protected function getVariantIdsIndexedByMainVariantIds(array $mainVariantsIds): array
     {
-        $result = $this->em->createQuery('SELECT IDENTITY(p.mainVariant) as id FROM ' . Product::class . ' p WHERE p.id IN (:productIds) AND p.variantType = :variantTypeVariant')
-            ->setParameter('productIds', $productIds)
-            ->setParameter('variantTypeVariant', Product::VARIANT_TYPE_VARIANT)
+        $query = 'SELECT 
+                DISTINCT CASE
+                    WHEN p.mainVariant IS NULL THEN p.id
+                    ELSE IDENTITY(p.mainVariant)
+                END AS main_variant_id,
+                p.id AS variant_id
+            FROM ' . Product::class . ' p
+            WHERE p.mainVariant IN (:productIds) OR p.id IN (:productIds)
+            ORDER BY main_variant_id, variant_id';
+
+        $result = $this->em->createQuery($query)
+            ->setParameter('productIds', $mainVariantsIds)
             ->getResult();
 
-        return array_column($result, 'id');
+        $groupedResults = [];
+
+        foreach ($result as $row) {
+            $groupedResults[$row['main_variant_id']][] = $row['variant_id'];
+        }
+
+        return $groupedResults;
     }
 
     /**
-     * @param int[] $productIds
+     * @param int[] $mainVariantsIds
      * @return int[]
      */
-    protected function dropVariantIdsFromProductIds(array $productIds): array
+    public function getIdsToRecalculate(array $mainVariantsIds): array
```

Comment:
Maybe I don't understand the code enough (and I truly don't understand it enough 😆 )
but wouldn't be this sql enough to get proper ids?

```
SELECT id FROM product WHERE id IN (/*main_variant_ids*/) OR main_variant_id IN (/*main_variant_ids*/)
```